From 625fd88e12ed00142bdc6d24ea29132ab569b315 Mon Sep 17 00:00:00 2001
From: Ren-Pei Zeng <kamesan@chromium.org>
Date: Mon, 17 Jul 2023 15:44:06 +0000
Subject: [PATCH] Revert "Merge
 ObjectProxy::CallMethodAndBlock{,WithErrorDetails}."

This introduces a temporary function to aid in migrating clients to the
new API.

---
 dbus/mock_object_proxy.h |  7 +++++++
 dbus/object_proxy.cc     | 33 +++++++++++++++++++++++++++++++++
 dbus/object_proxy.h      |  7 +++++++
 3 files changed, 47 insertions(+)

diff --git a/dbus/mock_object_proxy.h b/dbus/mock_object_proxy.h
index 9731ee5..ce762bf 100644
--- a/dbus/mock_object_proxy.h
+++ b/dbus/mock_object_proxy.h
@@ -28,6 +28,13 @@ class MockObjectProxy : public ObjectProxy {
       CallMethodAndBlock,
       base::expected<std::unique_ptr<Response>, Error>(MethodCall* method_call,
                                                        int timeout_ms));
+  MOCK_METHOD3(CallMethodAndBlockWithErrorDetails,
+               std::unique_ptr<Response>(MethodCall* method_call,
+                                         int timeout_ms,
+                                         Error* error));
+  MOCK_METHOD2(CallMethodAndBlockDeprecated,
+               std::unique_ptr<Response>(MethodCall* method_call,
+                                         int timeout_ms));
 
   // This method is not mockable because it takes a move-only argument. To work
   // around this, CallMethod() implementation here calls DoCallMethod() which is
diff --git a/dbus/object_proxy.cc b/dbus/object_proxy.cc
index 2e6798f..64e89f1 100644
--- a/dbus/object_proxy.cc
+++ b/dbus/object_proxy.cc
@@ -150,6 +150,39 @@ ObjectProxy::CallMethodAndBlock(MethodCall* method_call, int timeout_ms) {
   return result;
 }
 
+std::unique_ptr<Response> ObjectProxy::CallMethodAndBlockWithErrorDetails(
+    MethodCall* method_call,
+    int timeout_ms,
+    Error* error) {
+  bus_->AssertOnDBusThread();
+
+  if (!bus_->Connect() || !method_call->SetDestination(service_name_) ||
+      !method_call->SetPath(object_path_)) {
+    return nullptr;
+  }
+
+  // Send the message synchronously.
+  auto result =
+      bus_->SendWithReplyAndBlock(method_call->raw_message(), timeout_ms);
+  statistics::AddBlockingSentMethodCall(
+      service_name_, method_call->GetInterface(), method_call->GetMember());
+  if (!result.has_value()) {
+    LogMethodCallFailure(method_call->GetInterface(), method_call->GetMember(),
+                         result.error().name(), result.error().message());
+    *error = std::move(result.error());
+    return nullptr;
+  }
+
+  return std::move(result.value());
+}
+
+std::unique_ptr<Response> ObjectProxy::CallMethodAndBlockDeprecated(
+    MethodCall* method_call,
+    int timeout_ms) {
+  Error error;
+  return CallMethodAndBlockWithErrorDetails(method_call, timeout_ms, &error);
+}
+
 void ObjectProxy::CallMethod(MethodCall* method_call,
                              int timeout_ms,
                              ResponseCallback callback) {
diff --git a/dbus/object_proxy.h b/dbus/object_proxy.h
index 10d384f..c49be9c 100644
--- a/dbus/object_proxy.h
+++ b/dbus/object_proxy.h
@@ -116,6 +116,13 @@ class CHROME_DBUS_EXPORT ObjectProxy
   virtual base::expected<std::unique_ptr<Response>, Error> CallMethodAndBlock(
       MethodCall* method_call,
       int timeout_ms);
+  virtual std::unique_ptr<Response> CallMethodAndBlockWithErrorDetails(
+      MethodCall* method_call,
+      int timeout_ms,
+      Error* error);
+  virtual std::unique_ptr<Response> CallMethodAndBlockDeprecated(
+      MethodCall* method_call,
+      int timeout_ms);
 
   // Requests to call the method of the remote object.
   //
-- 
2.41.0.585.gd2178a4bd4-goog
